js/ui/searchbar.js

```diff
@@ -232,14 +244,54 @@ define([
       }
     },
 
+    searchHistory: function() {
+      var input = document.find(".searchbar").find(".search-box");
+      var history = [];
+      var currentIndex = 0;
+      var temporaryQuery = '';
+
+      return {
+        add: function(query) {
+          if (query != history[history.length - 1]) { // deduplicate
+            history.push(query);
+          }
+          currentIndex = history.length;
+          temporaryQuery = '';
+        },
+
+        showPrevious: function() {
+          if (currentIndex == history.length) {
+            temporaryQuery = input.value;
+            // skip previous search if we're already showing the same value
+            if (temporaryQuery == history[currentIndex-1]) {
+              currentIndex--;
+            }
+          }
+          if (currentIndex - 1 >= 0) {
+            input.value = history[--currentIndex];
+          }
+        },
+
+        showNext: function() {
+          if (currentIndex + 1 < history.length) {
+            input.value = history[++currentIndex];
+          } else if (currentIndex + 1 == history.length) {
+            currentIndex++;
+            input.value = temporaryQuery;
+          }
+        }
+      }
+    }(),
```

Comment:
I was thinking it should be encapsulated since it has its own variables that we don't want SearchBar to access directly. I still think it should be encapsulated, but created in the constructor instead of the prototype. 

I can change it to be just a part of SearchBar instead if you prefer though.